homeassistant/components/volkswagencarnet/__init__.py

```diff
@@ -0,0 +1,237 @@
+            {cv.slug: cv.string}),
+        vol.Optional(CONF_RESOURCES): vol.All(
+            cv.ensure_list, [vol.In(RESOURCES)])
+    }),
+}, extra = vol.ALLOW_EXTRA)
+
+def setup(hass, config):
+    """Setup Volkswagen Carnet component"""
+    interval = config[DOMAIN].get(CONF_SCAN_INTERVAL)
+    data = hass.data[DATA_KEY] = VolkswagenData(config)
+    
+    from volkswagencarnet import Connection
+
+    _LOGGER.debug("Creating connection to volkswagen carnet")
+    connection = Connection(
+        username = config[DOMAIN].get(CONF_USERNAME),
+        password = config[DOMAIN].get(CONF_PASSWORD),
+        #region = config[DOMAIN].get(CONF_REGION) # add support for this
+    )
+
+    # login to carnet
+    _LOGGER.debug("Logging in to volkswagen carnet")
+    connection._login()
+    if not connection.logged_in:
+        _LOGGER.warning('Could not login to volkswagen carnet, please check your credentials')
+
+    def is_enabled(attr):
+        """Return true if the user has enabled the resource."""
+        return attr in config[DOMAIN].get(CONF_RESOURCES, [attr])
+
+    def discover_vehicle(vehicle):
+        """Load relevant platforms."""
+        data.vehicles.add(vehicle.vin)
+        data.entities[vehicle.vin] = []
+
+        dashboard = vehicle.dashboard(mutable = config[DOMAIN][CONF_MUTABLE])
+
+        for instrument in (
+                instrument
+                for instrument in dashboard.instruments
+                if instrument.component in COMPONENTS and
+                is_enabled(instrument.slug_attr)):
+
+            data.instruments.add(instrument)
+            discovery.load_platform(hass, COMPONENTS[instrument.component], DOMAIN, (vehicle.vin,instrument.component,instrument.attr), config)
+
+    def update(now):
+        """Update status from Volkswagen Carnet"""
+        try:
+            # check if we can login again
+            if not connection.logged_in:
+                connection._login()
+                if not connection.logged_in:
+                    _LOGGER.warning('Could not login to volkswagen carnet, please check your credentials')
+                    return False
+            else:
+                if not connection.update(request_data = False):
+                    _LOGGER.warning("Could not query update from volkswagen carnet")
+                    return False
+                else:
+                    _LOGGER.debug("Updating data from volkswagen carnet")
+
+                    for vehicle in connection.vehicles:
+                        if vehicle.vin not in data.vehicles:
+                            _LOGGER.info("Adding data for VIN: %s from carnet" % vehicle.vin.lower())
+                            discover_vehicle(vehicle)
+                        
+                        for entity in data.entities[vehicle.vin]:
+                            entity.schedule_update_ha_state()
+
+                        dispatcher_send(hass, SIGNAL_STATE_UPDATED, vehicle)
```

Comment:
See above for an example how to connect the dispatch helper to a signal. Create one signal per VIN.